dev/bots/tool_subsharding.dart

```diff
@@ -38,42 +38,53 @@ class TestSpecs {
   }
 }
 
-/// Intended to parse the output file of `dart test --file-reporter json:file_name
-Map<int, TestSpecs> generateMetrics(File metrics) {
+class TestFileReporterResults {
   final Map<int, TestSpecs> allTestSpecs = <int, TestSpecs>{};
+  bool hasFailedTests = true;
+  final List<String> errors = <String>[];
+
+  void addTestSpec(Map<dynamic, dynamic> suite, int time) {
+    allTestSpecs[suite['id'] as int] = TestSpecs(
+      path: suite['path'] as String,
+      startTime: time,
+    );
+  }
+
+  void addMetricDone(int suiteID, int time) {
+    final TestSpecs testSpec = allTestSpecs[suiteID]!;
+    testSpec.endTime = time;
+  }
+
+  bool isMetricDone(Map<String, dynamic> entry) {
+    if (entry.containsKey('group') && entry['type'] as String == 'group') {
+      final Map<dynamic, dynamic> group = entry['group'] as Map<dynamic, dynamic>;
+      return allTestSpecs.containsKey(group['suiteID'] as int);
+    }
+    return false;
+  }
+}
+
+/// Intended to parse the output file of `dart test --file-reporter json:file_name
+TestFileReporterResults parseFileReporter(File metrics) {
```

Comment:
style nit: this function's whole purpose is to take input values, process them, and then return an initialized instance of a class--the paradigm with object-oriented programming is to do this from a constructor. Since you need to keep state as you're parsing, it makes sense to track this state in a factory, and then at the end, once you have all your state initialized, call a hidden private constructor from that factory, like:

```dart
class TestFileReporterResults {
  TestFileReporterResults._({
    required this.allTestSpecs,
    required this.hasFailedTests,
    required this.errors,
  });

  /// Intended to parse the output file of `dart test --file-reporter json:file_name
  factory TestFileReporterResults.fromFile(File metrics) {
    if (!metrics.existsSync()) {
      throw Exception('${metrics.path} does not exist');
    }

    final Map<int, TestSpecs> testSpecs = <int, TestSpecs>{};
    bool hasFailedTests = true;
    final List<String> errors = <String>[];

    for(final String metric in metrics.readAsLinesSync()) {
      final Map<String, dynamic> entry = json.decode(metric) as Map<String, dynamic>;
      if (entry.containsKey('suite')) {
        final Map<dynamic, dynamic> suite = entry['suite'] as Map<dynamic, dynamic>;
        addTestSpec(suite, entry['time'] as int, testSpecs);
      } else if (isMetricDone(entry, testSpecs)) {
        final Map<dynamic, dynamic> group = entry['group'] as Map<dynamic, dynamic>;
        final int suiteID = group['suiteID'] as int;
        addMetricDone(suiteID, entry['time'] as int, testSpecs);
      } else if (entry.containsKey('error') && entry.containsKey('stackTrace')) {
        errors.add('${entry['error']}\n ${entry['stackTrace']}');
      } else if (entry.containsKey('success') && entry['success'] == true) {
        hasFailedTests = false;
      }
    }

    return TestFileReporterResults._(allTestSpecs: testSpecs, hasFailedTests: hasFailedTests, errors: errors);
  }

  final Map<int, TestSpecs> allTestSpecs;
  bool hasFailedTests;
  final List<String> errors;

  static void addTestSpec(Map<dynamic, dynamic> suite, int time, Map<int, TestSpecs> allTestSpecs) {
    allTestSpecs[suite['id'] as int] = TestSpecs(
      path: suite['path'] as String,
      startTime: time,
    );
  }

  static void addMetricDone(int suiteID, int time, Map<int, TestSpecs> allTestSpecs) {
    final TestSpecs testSpec = allTestSpecs[suiteID]!;
    testSpec.endTime = time;
  }

  static bool isMetricDone(Map<String, dynamic> entry, Map<int, TestSpecs> allTestSpecs) {
    if (entry.containsKey('group') && entry['type'] as String == 'group') {
      final Map<dynamic, dynamic> group = entry['group'] as Map<dynamic, dynamic>;
      return allTestSpecs.containsKey(group['suiteID'] as int);
    }
    return false;
  }
}
```

Reply:
In addition, I can see you're typing the suite as `Map<dynamic, dynamic>`--can the key really be anything other than a string? Also, prefer `Object?` over `dynamic`, per https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-using-var-and-dynamic

Reply:
Sounds good; I wasn't able to find info on what types can be returned from `--file-reporter` thats why I decided to go with `Map<dynamic, dynamic>` rather than anything else, I can change this to `Object?` as per style

Reply:
If it's JSON, the spec only allows strings as the keys: https://www.json.org/json-en.html